TFNetworkLayer.py

```diff
@@ -9411,3 +9411,344 @@ def get_layer_class_name_list():
+    Output has shape (B,max(num_positions),None) and is zero-padded for each sequence.
+    Inputs: [LayerBase,LayerBase] [input_data,positions]: with
+    input_data: shape=(B,T,None)
+    positions: shape=(B,max(num_positions),) positions to gather for each sequence
+    from input_data (absolute positions within batch)
+    """
+    super(GatherPositionsLayer, self).__init__(**kwargs)
+    input_data = self.sources[0].output
+    positions = self.sources[1].output
+    batch_size = input_data.get_batch_dim()
+    seq_length = tf.shape(input_data.placeholder)[input_data.time_dim_axis]
+    feature_lengths = [tf.shape(input_data.placeholder)[ax] for ax in input_data.get_feature_batch_axes()]
+    data_flat = tf.reshape(input_data.placeholder, [batch_size * seq_length] + feature_lengths)
+
+    positions_w_offset = positions.copy_as_batch_major()
+    positions_w_offset.placeholder = tf.reshape(positions_w_offset.placeholder, [batch_size, -1])
+    offsets = tf.reshape(tf.range(0, batch_size, dtype=tf.int32), [batch_size, 1]) \
+              * seq_length * tf.ones_like(positions_w_offset.placeholder)
+    positions_w_offset.placeholder += offsets  # shift positions so they match the flattened representation
+    num_positions = positions_w_offset.get_sequence_lengths()
+    positions_flat = positions_w_offset.get_placeholder_flattened()
+    positions_flat = tf.reshape(positions_flat, shape=[-1, 1])
+    out_flat = tf.gather_nd(data_flat, positions_flat)  # gather the positions
+
+    # get data back to full shape
+    out_arr = tf.TensorArray(dtype=data_flat.dtype, size=0, dynamic_size=True, infer_shape=False)
+    i = tf.constant(0)
+    slice_start = tf.constant(0)
+
+    def while_condition(i, out_arr, slice_start):
+      return tf.less(i, batch_size)
+
+    def body(i, out_arr, slice_start):
+      # select what was gathered for each sequence:
+      out_arr = out_arr.write(i, out_flat[slice_start:slice_start + num_positions[i]])
+      return [tf.add(i, 1), out_arr, tf.add(slice_start, num_positions[i])]
+
+    _, out_arr, _ = tf.while_loop(while_condition, body, [i, out_arr, slice_start], back_prop=True)
+    self.output.placeholder = TFUtil.merge_tensor_array_with_padding(out_arr, num_positions, feature_lengths)
+    self.output.size_placeholder = {
+      0: num_positions
+    }
+
+  @classmethod
+  def get_out_data_from_opts(cls, sources, name, *args, **kwargs):
+    data = Data(
+      name=name,
+      shape=sources[0].output.shape,
+      dim=sources[0].output.dim,
+      sparse=False,
+      batch_dim_axis=sources[0].output.batch_dim_axis,
+      time_dim_axis=sources[0].output.time_dim_axis,
+      feature_dim_axis=sources[0].output.feature_dim_axis_or_unspecified,
+      dtype=sources[0].output.dtype)
+    return data
+
+
+class ProbabilisticMaskLayer(LayerBase):
+  """
+  This layer performs MLM (BERT-style) masking on the fly for a given source sequence.
+  """
+  layer_class = "probabilistic_mask"
```

Comment:
You should not add layers which are too specific (like this one).
If you want to write it that way, just copy that to your config, but this should not be part of Returnn. This is useful maybe only for a handful of experiments, and will likely be changed again and again, adding new options all the time, and becomes ugly and non-maintainable, until no-one will ever use it anymore. I explain that in the Returnn video lecture.

Reply:
I thought there was some demand in a layer like this (but I can see why you find this too specific). Also, I don't see a way to make this more generic and putting it in my config is just very ugly in combination with sisyphus. Therefore I'm going to keep this in my own fork for anyone to reproduce my MLM experiments.
Also, I will close this pull-request, since the rest is mainly useful in combination with `ProbabilisticMaskLayer`.

Reply:
It's not a good idea to keep your own fork around. Never. Everyone should always be able to reproduce any Returnn experiment with the current official Returnn master. (Or you say all these experiments are not interesting to anyone anymore, and no-one actually will care.)

I don't really know what problem there is with Sisyphus. This sounds like a very artificial problem, or not? If Sisyphus does not really allow this, you could also just put the layer code into a separate file, and then load that file in Returnn (either via `import`, or maybe just `eval`; see also `Util.custom_exec`).

But in any case, all what you do within this very specific layer was already possible before, or not? So you could have also written it in the config, using the existing layers. (Or what specifically was missing to allow this? See the Returnn video lecture. If any basic operation is missing, that is what a new layer is for.)